aya/src/programs/links.rs

```diff
@@ -121,38 +121,49 @@ impl FdLink {
     /// let pinned_link = fd_link.pin("/sys/fs/bpf/example")?;
     /// # Ok::<(), Error>(())
     /// ```
-    pub fn pin<P: AsRef<Path>>(mut self, path: P) -> Result<PinnedLink, PinError> {
-        let fd = self.fd.take().ok_or_else(|| PinError::NoFd {
-            name: "link".to_string(),
-        })?;
+    pub fn pin<P: AsRef<Path>>(self, path: P) -> Result<PinnedLink, PinError> {
         let path_string =
             CString::new(path.as_ref().to_string_lossy().into_owned()).map_err(|e| {
                 PinError::InvalidPinPath {
                     error: e.to_string(),
                 }
             })?;
-        bpf_pin_object(fd, &path_string).map_err(|(_, io_error)| PinError::SyscallError {
-            name: "BPF_OBJ_PIN".to_string(),
-            io_error,
+        bpf_pin_object(self.fd.unwrap(), &path_string).map_err(|(_, io_error)| {
+            PinError::SyscallError {
+                name: "BPF_OBJ_PIN".to_string(),
+                io_error,
+            }
         })?;
-        Ok(PinnedLink::new(PathBuf::from(path.as_ref()), fd))
+        Ok(PinnedLink::new(
+            PathBuf::from(path.as_ref()),
+            self.fd.unwrap(),
```

Comment:
That error could never happen: if you have a `FdLink` value, it means it hasn't been dropped, which means that the option is `Some`. Once `FdLink` is dropped, the option becomes None, and since the value is dropped by definition you can't call pin on it anymore.